service/mattermost/mattermost.go

```diff
@@ -0,0 +1,151 @@
+	return &Service{
+		setupLoginService(url, httpService),
+		httpService,
+		[]string{},
+	}
+}
+
+// LoginWithCredentials provides helper for authentication using Mattermost user/admin credentials.
+func (s *Service) LoginWithCredentials(ctx context.Context, loginID, password string) error {
+	// request login
+	if err := s.loginClient.Send(ctx, loginID, password); err != nil {
+		return errors.Wrapf(err, "failed login to Mattermost server")
+	}
+	return nil
+}
+
+// AddReceivers takes Mattermost channel IDs or Chat IDs and adds them to the internal channel ID list.
+// The Send method will send a given message to all these channels.
+func (s *Service) AddReceivers(channelIDs ...string) {
+	s.channelIDs = append(s.channelIDs, channelIDs...)
+}
+
+// Send takes a message subject and a message body and send them to added channel ids.
+// you will need a 'create_post' permission for your username.
+// refer https://api.mattermost.com/ for more info
+func (s *Service) Send(ctx context.Context, subject, message string) error {
+	for _, id := range s.channelIDs {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// create post
+			if err := s.messageClient.Send(ctx, id, subject+"\n"+message); err != nil {
+				return errors.Wrapf(err, "failed to send message")
+			}
+		}
+	}
+	return nil
+}
+
+// setups main message service for creating posts
+func setupMsgService(url string) *http.Service {
+	// create new http client for sending messages/notifications
+	httpService := http.New()
+
+	// add custom payload builder
+	httpService.AddReceivers(&http.Webhook{
+		URL:         url + "/api/v4/posts",
+		Header:      stdhttp.Header{},
+		ContentType: "application/json",
+		Method:      stdhttp.MethodPost,
+		BuildPayload: func(channelID, subjectAndMessage string) (payload any) {
+			return map[string]string{
+				"channel_id": channelID,
+				"message":    subjectAndMessage,
+			}
+		},
+	})
+
+	// add post-send hook for error checks
+	httpService.PostSend(func(req *stdhttp.Request, resp *stdhttp.Response) error {
+		if resp.StatusCode != stdhttp.StatusCreated {
+			b, _ := io.ReadAll(resp.Body)
+			return errors.New("failed to create post with status: " + resp.Status + " body: " + string(b))
+		}
+		return nil
+	})
+	return httpService
+}
+
+// setups login service to get token
+func setupLoginService(url string, msgService *http.Service) *http.Service {
+	// create another new http client for login request call.
+	httpService := http.New()
+
+	// append login path for the given mattermost server with custom payload builder.
+	httpService.AddReceivers(&http.Webhook{
+		URL:         url + "/api/v4/users/login",
+		Header:      stdhttp.Header{},
+		ContentType: "application/json",
+		Method:      stdhttp.MethodPost,
+		BuildPayload: func(loginID, password string) (payload any) {
+			return map[string]string{
+				"login_id": loginID,
+				"password": password,
+			}
+		},
+	})
+
+	// Add pre-send hook to log the request before it is sent.
```

Comment:
I can see the logic here by adding the presend log, but generally we want to avoid logging. If you would like this presend hook can be added to examples in the README and doc.go, but we shouldn't assume this is what the user means to do.

Reply:
Makes sense, to provide users the ability to add these hooks I thought of two options:
1. Expose loginClient and messageClient of `Service` struct so that users can use these to attach hooks.
2. Provide these hooks as a method for mattermost service itself which users can just call PreSend and PostSend as `Send`.

I just went with 2nd option as it suits the current style that all other services have(doesn't expose `client` to users). one downside for this is to have separate methods for login and message clients which I thought might make look complex for simple service, so just added callable hooks for message client in which user might be more interested in.

Also removed the PreSend and PostSend logging from the login client as suggested.

Reply:
Hey @shivuslr41
The approach you took is fine in my eyes. I don't think it would be necessary (generally) to add hooks for the loginClient. I'm trying to think of use cases where hooks would be useful for mattermost login but can't think of any. Have any ideas here?

Reply:
Same here, couldn't think of any better use case apart from simple log messages on successful login. I think we can have hooks only for message clients now until some better use case comes up!